### PR TITLE
DCUNet and DCCRNet

### PR DESCRIPTION
**Pull request by jonashaag**
This is a WIP implementation of [Phase-aware Single-stage Speech Denoising and Dereverberation with U-Net](https://arxiv.org/abs/2006.00687).

I also want to implement [DCCRN: Deep Complex Convolution Recurrent Network for Phase-Aware Speech Enhancement](https://arxiv.org/abs/2008.00264) soon.

I'd like to get feedback on

- Usage of PyTorch 1.7+ native complex128 dtype rather than Asteroid style complex type or torchaudio style complex type.
- Code structure, class names, code placing. Many things I wasn't sure where to place in the Asteroid codebase.
- Location of complex norms, activations, etc.
- Dealing with input shapes in `DCUMaskNet.forward`. Due to the encoder/decoder structure only specific T/F dimensions will work without padding or other workarounds.

### CODE REVIEW

**Comment by jonashaag** [comment]
Should be good to review now. Sorry for making so many changes in a single PR. If it is difficult to review, I will split.

Probably misses lots of tests; I will wait for codecov to tell me the uncovered places and add some tests.

File: asteroid/models/base_models.py [hunk 1]
```diff
@@ -42,10 +51,9 @@ def separate(self, wav, save_dir=None, force_overwrite=False, **kwargs):
         was_file = False
         if isinstance(wav, str):
             import soundfile as sf
-
             was_file = True
             filename = wav
-            wav, fs = sf.read(wav, dtype="float32")
+            wav, fs = sf.read(wav, dtype='float32')
```
Comment (jonashaag) [inline asteroid/models/base_models.py:56]: Oops, what happened here?

File: asteroid/utils/generic_utils.py [hunk 2]
```diff
@@ -75,12 +75,34 @@ def get_wav_random_start_stop(signal_len, desired_len=4 * 8000):
     Returns:
         tuple: random start integer, stop integer.
     """
-    if signal_len == desired_len or desired_len is None:
```
Comment (jonashaag) [inline asteroid/utils/generic_utils.py:78]: Fixed:

- Edge case `desired_len > signal_len`
- Always return stop as int (as documented)

File: tests/models/models_test.py [hunk 3]
```diff
@@ -60,54 +70,52 @@ def test_dprnntasnet_sep():
 
 @pytest.mark.parametrize("fb", ["free", "stft", "analytic_free", "param_sinc"])
 def test_save_and_load_dprnn(fb):
-    model1 = DPRNNTasNet(
```
Comment (jonashaag) [inline tests/models/models_test.py:63]: Deduplicated some tests

File: asteroid/dsp/overlap_add.py [hunk 4]
```diff
@@ -68,17 +68,17 @@ def ola_forward(self, x):
         batch, channels, n_frames = x.size()
         # Overlap and add:
         # [batch, chans, n_frames] -> [batch, chans, win_size, n_chunks]
-        folded = torch.nn.functional.unfold(
```
Comment (jonashaag) [inline asteroid/dsp/overlap_add.py:71]: Fixed name

File: asteroid/masknn/convolutional.py [hunk 5]
```diff
@@ -383,6 +388,112 @@ def get_config(self):
...
+        self.norm = norms.get_complex(norm_type)(out_chan)
+
+        activation_class = activations.get_complex(activation)
+        self.activation = activation_class()
+
+    def forward(self, x: complex_nn.ComplexTensor):
+        return self.activation(self.norm(self.deconv(x)))
+
+
+class DCUMaskNet(BaseDCUMaskNet):
```
Comment (mpariente) [inline asteroid/masknn/convolutional.py:485]: Maybe document here that there is the `default_architecture` classmethod? With valid architecture names? 
Or add the `show_available_architectures` class method in the base class? 
Otherwise, I don't see how the user discovers which architecture to use.

File: asteroid/models/base_models.py [hunk 6]
```diff
@@ -193,10 +203,7 @@ def __init__(self, encoder, masker, decoder, encoder_activation=None):
         self.decoder = decoder
 
         self.encoder_activation = encoder_activation
-        if encoder_activation:
-            self.enc_activation = activations.get(encoder_activation)()
-        else:
-            self.enc_activation = activations.get("linear")()
+        self.enc_activation = activations.get(encoder_activation or "linear")()
```
Comment (mpariente) [inline asteroid/models/base_models.py:206]: That's pretty, I never use it.

File: asteroid/models/base_models.py [hunk 7]
```diff
@@ -208,34 +215,98 @@ def forward(self, wav):
...
+        (output of the decoder) before original shape reconstruction.
+
+        Args:
+            decoded (Tensor of shape (batch, n_src, time)):
+                Output of the decoder, before original shape reconstruction.
+
+        Return:
+            Transformed `decoded`
+        """
+        return decoded
```
Comment (mpariente) [inline asteroid/models/base_models.py:292]: I love this ! 
@popcornell do you feel like to give your opinion on this as well?

File: asteroid/models/base_models.py [hunk 8]
```diff
@@ -177,13 +185,15 @@ def get_model_args(self):
         raise NotImplementedError
 
 
-class BaseTasNet(BaseModel):
+class BaseEMDNet(BaseModel):
```
Comment (mpariente) [inline asteroid/models/base_models.py:188]: @popcornell Opinion about renaming that to `BaseEMDNet`? 
(with code to make it backward compatible, obviously)

File: asteroid/models/base_models.py [hunk 9]
```diff
@@ -5,8 +5,17 @@
 import numpy as np
 
 from .. import torch_utils
-from ..utils.hub_utils import cached_download
 from ..masknn import activations
+from ..utils.hub_utils import cached_download
```
Comment (mpariente) [inline asteroid/models/base_models.py:9]: Can we have this whole file in a separate PR please? 
This is quite important and I'd like to be able to dissociate from the rest. 
Thanks

File: asteroid/models/dccrnet.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        stft_kernel_size (int): STFT frame length to use
+        stft_stride (int, optional): STFT hop length to use.
+
+    References:
+        [1] : "DCCRN: Deep Complex Convolution Recurrent Network for Phase-Aware Speech Enhancement",
+        Yanxin Hu et al.
+        https://arxiv.org/abs/2008.00264
+    """
+
+    def __init__(self, architecture, stft_kernel_size=512, stft_stride=None):
```
Comment (mpariente) [inline asteroid/models/dccrnet.py:23]: Can we put the default architecture to `"DCCRN-CL"`?

**Review by mpariente — COMMENTED** [review]
This is a great PR ! 
I love the new BaseTasNet with the hooks, it's very nice. 
There are few things to address/discuss in the comments but this is in a very good shape

**Comment by jonashaag** [comment]
I split the changes into multiple commits, but did not split into multiple PRs, so as to not lose your review comments.

File: asteroid/complex_nn.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    @functools.wraps(f)
+    def cf(x):
+        return as_torch_complex((f(x.real), f(x.imag)))
+
+    cf.__name__ == f"{f.__name__} (complex)"
+    cf.__qualname__ == f"{f.__qualname__} (complex)"
+    return cf
+
+
+class OnComponents(nn.Module):
```
[the remark this answers is not in the document]
Comment (jonashaag) [reply asteroid/complex_nn.py:90]: It is different: Your code uses only a single instance of `m` and thus only 1x the weights

File: asteroid/complex_nn.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    def __init__(self, module_cls, *args, **kwargs):
+        super().__init__()
+        self.re_module = module_cls(*args, **kwargs)
+        self.im_module = module_cls(*args, **kwargs)
+
+    def forward(self, x):
+        return as_torch_complex((self.re_module(x.real), self.im_module(x.imag)))
+
+
+class ComplexMultiplicationWrapper(nn.Module):
```
[the remark this answers is not in the document]
Comment (jonashaag) [reply asteroid/complex_nn.py:107]: I know... do you have an idea how to call it?

File: asteroid/complex_nn.py [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+
+
+def bound_complex_mask(mask: ComplexTensor, bound_type="tanh"):
+    r"""Bound a complex mask, as proposed in [1], section 3.2.
+
+    Valid bound types, for a complex mask $M = |M| \cdot e^{i \varphi(M)}$:
+
+    - Unbounded ("UBD"): $M_{\mathrm{UBD}} = M$
+    - Sigmoid ("BDSS"): $M_{\mathrm{BDSS}} = \sigma(|M|) e^{i \sigma(\varphi(M))}$
+    - Tanh ("BDT"): $M_{\mathrm{BDT}} = \mathrm{tanh}(|M|) e^{i \varphi(M)}$
```
Comment (jonashaag) [inline asteroid/complex_nn.py:151]: Added formulas, is this what you wanted? sqrt(2) is a possible magnitude value for BDSS.

File: asteroid/filterbanks/transforms.py [hunk 14]
```diff
@@ -241,6 +284,7 @@ def to_torchaudio(tensor, dim=-2):
         :class:`torch.Tensor`:
             torchaudio-style complex-like torch tensor.
     """
+    # TODO: dim must always be -1
```
[the remark this answers is not in the document]
Comment (jonashaag) [reply asteroid/filterbanks/transforms.py:288]: Oversight. Will remove.

File: asteroid/masknn/base.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (jonashaag) [reply asteroid/masknn/base.py:1]: I dunno, I just needed a place to stick these :-D

File: asteroid/masknn/convolutional.py (hunk 5 above)
Comment (jonashaag) [reply asteroid/masknn/convolutional.py:483]: This is now documented here.

File: asteroid/masknn/recurrent.py [hunk 16]
```diff
@@ -1,10 +1,16 @@
...
 from torch.nn.functional import fold, unfold
 
-from . import norms, activations
-from .norms import GlobLN, CumLN
+from .. import complex_nn
 from ..utils import has_arg
+from . import activations, norms
+from ._dccrn_architectures import DCCRN_ARCHITECTURES
+from .base import BaseDCUMaskNet
+from .norms import CumLN, GlobLN
```
Comment (jonashaag) [inline asteroid/masknn/recurrent.py:13]: Umm... something is wrong with black or isort here.

File: asteroid/models/base_models.py [hunk 17]
```diff
@@ -99,7 +109,7 @@ def file_separate(
             sf.write(save_name, est_src, fs)
 
     def _separate(self, wav, *args, **kwargs):
-        """ Hidden separation method
```
Comment (jonashaag) [inline asteroid/models/base_models.py:102]: Did not mean to commit this... but I wonder why it hasn't already been formatted with black?

File: asteroid/models/dcunet.py [hunk 18]
```diff
@@ -0,0 +1,62 @@
+import torch
+
+from .. import complex_nn
+from ..filterbanks import make_enc_dec
+from ..filterbanks.transforms import from_torchaudio
+from ..masknn.convolutional import DCUMaskNet
+from .base_models import BaseEMDNet
+
+
+class BaseDCUNet(BaseEMDNet):
```
Comment (jonashaag) [inline asteroid/models/dcunet.py:10]: Moved common code from DCU and DCCRN here

File: asteroid/utils/generic_utils.py [hunk 19]
```diff
@@ -75,12 +75,34 @@ def get_wav_random_start_stop(signal_len, desired_len=4 * 8000):
...
-        stop = None
-    else:
-        stop = rand_start + desired_len
+        return 0, signal_len
+    rand_start = np.random.randint(0, max(1, signal_len - desired_len))
+    stop = min(signal_len, rand_start + desired_len)
     return rand_start, stop
+
+
+def unet_decoder_args(encoders, *, skip_connections):
```
[the remark this answers is not in the document]
Comment (jonashaag) [reply asteroid/utils/generic_utils.py:85]: Maybe it's specific, I don't know... I have seen it many times, e.g. "Singing Voice Separation with Deep U-Net Convolutional Networks", Spleeter, DCUNet, DCCRNet, ... granted, those are all variations of the same u-net structure. But I think even in the original u-net paper they use this structure.

`skip_connections`: The `*` forces the following arguments to be keyword-only arguments. It's just a way to force people to call it like `unet_decoder_args(..., skip_connections=True)` rather than `unet_decoder_args(..., True)`, which is less readable code. So just a way to force people to write better code. I always do it for arguments that don't have an obvious meaning at the caller site. But I don't have strong opinions about it, so if you don't like it I can remove it.

File: asteroid/utils/generic_utils.py [hunk 20]
```diff
@@ -75,12 +75,34 @@ def get_wav_random_start_stop(signal_len, desired_len=4 * 8000):
...
+            Arguments to be used to construct decoders
+    """
+    decoder_args = []
+    for enc_in_chan, enc_out_chan, enc_kernel_size, enc_stride, enc_padding in reversed(encoders):
+        if skip_connections and decoder_args:
+            skip_in_chan = enc_out_chan
+        else:
+            skip_in_chan = 0
+        decoder_args.append(
+            (enc_out_chan + skip_in_chan, enc_in_chan, enc_kernel_size, enc_stride, enc_padding)
```
[the remark this answers is not in the document]
Comment (jonashaag) [reply asteroid/utils/generic_utils.py:106]: Yes!

**Comment by jonashaag** [comment]
I updated black to latest version because it now also formats docstrings.

**Comment by mpariente** [comment]
I didn't comment on much on the `BaseTasNet` part on purpose. 
I'd like to cherry-pick that commit (88701900175535548383dc75ced135f12d81250b) to make a new PR, this is more visible for other people which might be interested, and for future reference. For the rest, it can stay like that, thanks a lot!

File: asteroid/utils/generic_utils.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (mpariente) [reply asteroid/utils/generic_utils.py:85]: - Ok to keep this code in utils then, thanks.
- Ok for that. I would maybe add `True` as a default value, WDYT?

File: asteroid/complex_nn.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mpariente) [reply asteroid/complex_nn.py:90]: It can perform (documented) copy of the instance right? 
But maybe you're right that the from the first way, it's clearer that two modules are instantiated.

File: asteroid/data/avspeech_dataset.py [hunk 21]
```diff
@@ -36,16 +36,16 @@ def get_frames(video):
...
-            video_path (str,Path): Path to video (mp4).
-            audio_path (str,Path): Path to audio (wav).
-            embed_dir (str,Path): Path to directory that stores embeddings.
-            sr (int): sampling rate of audio.
-            video_start_length: video part no. [1]
-            fps (int): fps of video.
-            signal_len (int): length of the signal
-
-        .. note:: each video consists of multiple parts which consists of fps*signal_len frames.
+    Args:
```
Comment (mpariente) [inline asteroid/data/avspeech_dataset.py:39]: All black related commit should come in a single PR. 
I'm sorry to be rigid on that but this will make a cleaner history. 
Imagine if we introduced a bug with this PR and we have the revert everything and check what made the bug, and redo all the formatting. Having separate PRs is better.

**Review by mpariente — COMMENTED** [review]
Few feedback for now but I stop to review because it's not clear what there is to review. 
I'll keep reviewing when there will be the 3 PRs: this one, the model rewriting which is cool and the black formatting. Sorry about that.

**Comment by sufeidechabei** [comment]
Hello, guys, does the the result of DCCRN over all the datasets match with the original paper？

**Comment by sufeidechabei** [comment]
@jonashaag

**Comment by jonashaag** [comment]
Haven't done a full training yet, so I don't know!

**Comment by mpariente** [comment]
Hey @jonashaag, can we split this PR into three now please?

**Comment by jonashaag** [comment]
Yes will do it next week.

File: asteroid/complex_nn.py [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+            return torch.view_as_complex(x)
+        elif is_asteroid_complex:
+            return torch.view_as_complex(transforms.to_torchaudio(x, asteroid_dim))
+        else:
+            raise ValueError(
+                f"Do not know how to convert tensor of shape {x.shape}, dtype={x.dtype} to complex"
+            )
+
+
+def on_components(f):
```
[the remark this answers is not in the document]
Comment (jonashaag) [reply asteroid/complex_nn.py:69]: I like it!

**Comment by jonashaag** [comment]
Removed all changes not related to DCUNet. We can continue discuss DCUNet changes here.

File: asteroid/complex_nn.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (mpariente) [reply asteroid/complex_nn.py:107]: Still need to solve this.

File: asteroid/filterbanks/transforms.py [hunk 23]
```diff
@@ -242,6 +285,7 @@ def to_torchaudio(tensor, dim=-2):
...
```
Comment (mpariente) [inline asteroid/filterbanks/transforms.py:288]:
```suggestion
```

File: asteroid/masknn/recurrent.py [hunk 24]
```diff
@@ -484,3 +493,81 @@ def get_config(self):
...
+        [1] : "DCCRN: Deep Complex Convolution Recurrent Network for Phase-Aware Speech Enhancement",
+        Yanxin Hu et al.
+        https://arxiv.org/abs/2008.00264
+    """
+
+    def __init__(self, in_size, hid_size=128, rnn_type="LSTM", norm_type=None):
+        super().__init__()
+
+        self.rnn = ComplexSingleRNN(rnn_type, in_size, hid_size)
+        self.linear = complex_nn.ComplexMultiplicationWrapper(nn.Linear, hid_size, in_size)
```
Comment (mpariente) [inline asteroid/masknn/recurrent.py:521]: Why the RNN as `ComplexSingleRNN` and the linear layer not? Just wondering.

File: asteroid/models/base_models.py [hunk 25]
```diff
@@ -195,13 +206,15 @@ def get_model_args(self):
         raise NotImplementedError
 
 
-class BaseTasNet(BaseModel):
+class BaseEMDNet(BaseModel):
```
Comment (mpariente) [inline asteroid/models/base_models.py:209]: Let's fix the conflicts with master on this.

File: asteroid/models/dcunet.py [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+
+    def postprocess_masked(self, masked_tf_rep):
+        return from_torchaudio(torch.view_as_real(masked_tf_rep))
+
+    def get_model_args(self):
+        """Arguments needed to re-instantiate the model."""
+        model_args = {
+            "architecture": self.architecture,
+            "stft_kernel_size": self.stft_kernel_size,
+            "stft_stride": self.stft_stride,
```
Comment (mpariente) [inline asteroid/models/dcunet.py:42]: What about `masknet_kwargs`?

**Review by mpariente — COMMENTED** [review]
This is great ! 
Let's 
- Converge on `ComplexMultiplicationWrapper`
- Solve the conflicts with master 
- And merge ! :rocket: 

We'll make a new release after that!

File: asteroid/masknn/recurrent.py (hunk 24 above)
Comment (jonashaag) [reply asteroid/masknn/recurrent.py]: Good point, changed it.

**Comment by jonashaag** [comment]
I *think* I have addressed all review comments

**Comment by mpariente** [comment]
This looks great, thanks a lot again ! 
The CI fails though..

**Comment by jonashaag** [comment]
Ah well, multiple places with recursive imports... this "let's flatten the namespaces by importing things in the parent module" thing causes lots of problems. (For example importing models to `asteroid.models.` or even `asteroid.`)

**Comment by mpariente** [comment]
> Ah well, multiple places with recursive imports... this "let's flatten the namespaces by importing things in the parent module" thing causes lots of problems. (For example importing models to asteroid.models. or even asteroid.) 

I think models should be importable from `asteroid` directly. For `torch_utils` it was for backwards compat because there was a `torch_utils.py` file before that. 

Thanks for fixing it. Now there is a new "real" failure.

**Comment by mpariente** [comment]
That was a nice one, thanks a lot Jonas !
